cpp/include/cudf/detail/cuco_helpers.hpp

```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/mr/device/polymorphic_allocator.hpp>
+
+namespace cudf::detail {
+
+class cuco_allocator
```

Comment:
Please add docs explaining why do we need this allocator, and why we need to subclass this rmm allocator for it? Why don't we use other allocator?

Reply:
> why do we need this allocator

By removing `::default_allocator`, `polymorphic_allocator` becomes the only option we have for cuco data structures. 

> Why don't we use other allocator?

What are the alternatives?

> why we need to subclass this rmm allocator for it?

This is mainly for the convenience of users. I started with just a type alias like:
```cpp
using cuco_allocator =  rmm::mr::stream_allocator_adaptor<rmm::mr::polymorphic_allocator<char>>;
```
Then to construct this allocator, users need to pass the `polymorphic_allocator` manually as well:
```cpp
auto alloc = cuco_allocator{rmm::mr::polymorphic_allocator<char>{}, stream};
```
Now with inheritance, constructing an allocator can be as simple as:
```cpp
auto alloc = cuco_allocator{stream};
```

Another option is composition where we make `rmm::mr::stream_allocator_adaptor<rmm::mr::polymorphic_allocator<char>>` as a data member of `cuco_allocator`. In that case, we have to "facade" all public member functions of `stream_allocator_adaptor` which I find laborious e.g.:
```cpp
class cuco_allocator {
  public:
    cuco_allocator(... stream) : _allocator{rmm::mr::polymorphic_allocator<char>{}, stream} {}

    value_type* allocate(...) {this->_allocator.allocate(...);}
    void deallocate(...) {this->_allocator.deallocate(...);}
    ...
  private:
    rmm::mr::stream_allocator_adaptor<rmm::mr::polymorphic_allocator<char>> _allocator;
};
```

With that, they are the implementation details I tend not to include in the doc.

Reply:
Then a similar question is, why don't we implement this default allocator in cuco? By doing so, other down stream application can use it too, not just in cudf.

Reply:
`cuco` doesn't depend on `rmm`

Reply:
> With that, they are the implementation details I tend not to include in the doc.

I still think we need a short description (without much that detail) for the class. Others unfamiliar with the code may still want to ask why do they need such allocator instead of the default allocator used in `cuco::` class constructors. For example:
```
@brief An adaptor to rmm allocator used for cuco data structures.
```